File: Basic/Chapter 5/acwing2.go
package main

import (
	"fmt"
)

const maxn int = 1010

var (
	w, v, f [maxn]int
)

func max(a, b int) int {
	if a < b {
		return b
	} else {
		return a
	}
}

func main() {
	var n, m int
	fmt.Scanf("%d%d", &n, &m)
	for i := 1; i <= n; i++ {
		fmt.Scanf("%d%d", &v[i], &w[i])
	}
	for i := 1; i <= n; i++ {
		for j := m; j >=0 ; j-- {
			if j >= v[i] {
				f[j] = max(f[j], f[j-v[i]]+w[i])
			}
		}
	}
	fmt.Println(f[m])
}

File: acm/acwing2.cpp
#include <iostream>
using namespace std;
const int maxn=1010;
const int inf=1<<30;

int f[maxn],v[maxn],w[maxn],n,m;

int main(){
	while(cin>>n>>m){
		for(int i=1;i<maxn;i++)
			f[i]=-inf;
		for(int i=1;i<=n;i++)
			cin>>v[i]>>w[i];
		for(int i=1;i<=n;i++){
			for(int j=m;j>=v[i];j--)
				f[j]=max(f[j],f[j-v[i]]+w[i]);
		}
		cout << f[m] << endl;
	}
	return 0;
}